satellite/metainfo/service.go

```diff
@@ -106,6 +112,10 @@ func (s *Service) UpdatePieces(ctx context.Context, path string, ref *pb.Pointer
 			if exists {
 				return nil, Error.New("piece to add already exists (piece no: %d)", piece.PieceNum)
 			}
+			_, exists = uniqueNodeMap[piece.NodeId]
+			if exists {
+				return nil, ErrDuplicatedNodeID.New("piece num: %d, node id: %s", piece.PieceNum, piece.NodeId.String())
+			}
 			pieceMap[piece.PieceNum] = piece
```

Comment:
Might also be worth doing `delete(uniqueNodeMap, piece.NodeId)` in the `for ... toRemove` loop above. I don't think it will ever be the case that we remove and add the same node ID in the same `UpdatePieces` request, but we should decide if this situation should cause an `ErrDuplicatedNodeID` issue.

Reply:
Hmmm, I think technically, a node could store a different piece for the same segment. But I don't know when that would be the case

Reply:
I don't think it would ever happen, I guess it is just a matter of whether we want to support it technically. I will approve the change either way.

Reply:
That potentially could happen when two different systems race. e.g. repair and graceful exit at the same time; or some other combination, if that doesn't happen. Either way, it'll be safer to assume that it might.

Reply:
So does that mean we *should not* return a duplicate node ID error if the same node is being added and removed in one request?

Reply:
Ideally we shouldn't have a duplicate, but there's no guarantee that we currently don't have one. Making this change requires verifying the whole database to not contain it, otherwise changes like these could break repair for those pointers.

Reply:
Ok so maybe it's best for us to move this check to the graceful exit code for now.